lib/src/util/side_effect_visitor.dart

```diff
@@ -0,0 +1,546 @@
+  }
+
+  @override
+  visitNativeFunctionBody(NativeFunctionBody node) {
+    super.visitNativeFunctionBody(node);
+  }
+
+  @override
+  visitNullLiteral(NullLiteral node) {
+    super.visitNullLiteral(node);
+  }
+
+  @override
+  visitParenthesizedExpression(ParenthesizedExpression node) {
+    super.visitParenthesizedExpression(node);
+  }
+
+  @override
+  visitPartDirective(PartDirective node) {
+    super.visitPartDirective(node);
+  }
+
+  @override
+  visitPartOfDirective(PartOfDirective node) {
+    super.visitPartOfDirective(node);
+  }
+
+  @override
+  visitPostfixExpression(PostfixExpression node) {
+    _hasSideEffect = true;
+  }
+
+  @override
+  visitPrefixedIdentifier(PrefixedIdentifier node) {
+    super.visitPrefixedIdentifier(node);
+  }
+
+  @override
+  visitPrefixExpression(PrefixExpression node) {
+    _hasSideEffect = true;
+  }
+
+  @override
+  visitPropertyAccess(PropertyAccess node) {
+    super.visitPropertyAccess(node);
+  }
+
+  @override
+  visitRedirectingConstructorInvocation(RedirectingConstructorInvocation node) {
+    _hasSideEffect = true;
+  }
+
+  @override
+  visitRethrowExpression(RethrowExpression node) {
+    super.visitRethrowExpression(node);
+  }
+
+  @override
+  visitReturnStatement(ReturnStatement node) {
+    super.visitReturnStatement(node);
+  }
+
+  @override
+  visitScriptTag(ScriptTag node) {
+    super.visitScriptTag(node);
+  }
+
+  @override
+  visitShowCombinator(ShowCombinator node) {
+    super.visitShowCombinator(node);
+  }
+
+  @override
+  visitSimpleFormalParameter(SimpleFormalParameter node) {
+    super.visitSimpleFormalParameter(node);
+  }
+
+  @override
+  visitSimpleIdentifier(SimpleIdentifier node) {
+    super.visitSimpleIdentifier(node);
+  }
+
+  @override
+  visitSimpleStringLiteral(SimpleStringLiteral node) {
+    super.visitSimpleStringLiteral(node);
+  }
+
+  @override
+  visitStringInterpolation(StringInterpolation node) {
+    super.visitStringInterpolation(node);
+  }
+
+  @override
+  visitSuperConstructorInvocation(SuperConstructorInvocation node) {
+    super.visitSuperConstructorInvocation(node);
+  }
+
+  @override
+  visitSuperExpression(SuperExpression node) {
+    super.visitSuperExpression(node);
+  }
+
+  @override
+  visitSwitchCase(SwitchCase node) {
+    super.visitSwitchCase(node);
+  }
+
+  @override
+  visitSwitchDefault(SwitchDefault node) {
+    super.visitSwitchDefault(node);
+  }
+
+  @override
+  visitSwitchStatement(SwitchStatement node) {
+    super.visitSwitchStatement(node);
+  }
+
+  @override
+  visitSymbolLiteral(SymbolLiteral node) {
+    super.visitSymbolLiteral(node);
+  }
+
+  @override
+  visitThisExpression(ThisExpression node) {
+    super.visitThisExpression(node);
+  }
+
+  @override
+  visitThrowExpression(ThrowExpression node) {
+    super.visitThrowExpression(node);
+  }
+
+  @override
+  visitTopLevelVariableDeclaration(TopLevelVariableDeclaration node) {
+    super.visitTopLevelVariableDeclaration(node);
```

Comment:
I don't think you'll ever visit one of these, but they can have a side-effect if there is an initializer and the variable is referenced somewhere that could be executed.

Or we could view the reference to the variable as potentially having a side effect if this is the first time the variable is referenced. (But that would mean looking at all identifiers to see whether they refer to a static field or top-level variable.)